src/instrumenter/instrument.ts

```diff
@@ -265,8 +284,18 @@ export function flattenExpr(expr: SNode, ctx: TranspilingContext): [SNode, SBind
     if (expr instanceof SId) {
         // Case when the id is a let variable (e.g. `x` in `let x := 1 in x` or `y` in `let y, z := foo() in y+z`)
         if (expr.defSite instanceof Array) {
-            const field = ctx.getLetBinding(expr.defSite);
-            return [getTmpVar(field, expr, expr.src), []];
+            const [defNode, idx] = expr.defSite;
+
+            if (defNode instanceof SLet) {
+                const field = ctx.getLetBinding([defNode, idx]);
+                return [getTmpVar(field, expr, expr.src), []];
+            } else if (defNode instanceof SUserFunctionDefinition) {
+                const renamedId = new SId(ctx.getUserFunArg(defNode, idx), expr.src);
+                renamedId.defSite = expr.defSite;
+                return [_registerNode(renamedId, expr), []];
+            } else {
+                throw new Error(`Unknown array def site`);
+            }
```

Comment:
Can we reduce nesting here by using early `return`s?

Reply:
Fixed in 031fe1c.